tools/mksurfdata_map/src/mksurfdat.F90

```diff
@@ -1399,13 +1423,28 @@ subroutine normalizencheck_landuse(ldomain)
           call abort()
        end if
 
+       ! IV: adjust preconditions:
+       ! If pctwet + pcturb + pctgla + pctlak > 100: pct lak is adjusted so that total is 100
+       ! pctwet + pcturb + pctgla cannot be >100
+       ! TO DO: adjust in subroutine description if added. 
+       
        suma = pctlak(n) + pctwet(n) + pcturb(n) + pctgla(n)
-       if (suma > (100._r8 + tol_loose)) then
-          write(6,*) subname, ' ERROR: pctlak + pctwet + pcturb + pctgla must be'
-          write(6,*) '<= 100% before calling this subroutine'
-          write(6,*) 'n, pctlak, pctwet, pcturb, pctgla = ', &
-               n, pctlak(n), pctwet(n), pcturb(n), pctgla(n)
-          call abort()
+       if ( suma > (100._r8 + tol_loose) ) then
+          
+          ! calc pct lake as to fill cell
+          pctlak(n) = 100._r8 - (pctwet(n) + pcturb(n) + pctgla(n))
+          
+          ! recalculate sum
+          suma = pctlak(n) + pctwet(n) + pcturb(n) + pctgla(n)
+          
+          if  (suma > (100._r8 + tol_loose)) then
+            write(6,*) subname, ' ERROR: pctwet + pcturb + pctgla must be'
+            write(6,*) '<= 100% before calling this subroutine'
+            write(6,*) 'n, pctlak, pctwet, pcturb, pctgla = ', &
+                n, pctwet(n), pcturb(n), pctgla(n)
+            call abort()
+          end if
```

Comment:
1. I included the check on the sum of all land cover types not exceeding > 100 % within the time loop (55a0fb480a67059d36de25eecec80a731a021b57). However, different to the code for the surface dataset, I kept that when special land units > 100%,  pct_lak is adjusted to fill up the cell to 100%., instead of subtracting the excess from the most dominant land cover. To me this makes most sense, as the other special land units are constant in time. Do you agree? 

2. About rounding to the nearest integer: I didn't add this. I don't have strong opinions on this, and don't really know why this would be necessary. But if so, I could still add it.

Reply:
Thanks, @Ivanderkelen . I'm thinking that it may be important to be consistent in the logic for the surface dataset and the landuse_timeseries file – otherwise, we could get different behavior from using a landuse timeseries file even if the lake area is supposed to be the same there as in the surface dataset. Also, other landunits like urban may become dynamic in the future.

For similar reasons, I'm now leaning more towards thinking this should be consistent in terms of the rounding to nearest integer. We could change this rounding code if it's important to capture lake areas < 0.5% on the grid cell, or to capture more precision in the lake area dynamics, but if we do change this for the landuse timeseries file, I'm still thinking that we should be consistent in the surface dataset, too.

However, I'm also realizing that, for the sake of consistency and correctness – especially with the treatment of wetland % – we may actually need slightly more extensive changes here. In particular, I think it may be important in some cases that we reset pct area of the various special landunits to their pre-corrected values, and then redo the various landunit adjustments from the beginning. This way, we'll be doing the same adjustments from scratch each year, rather than (for example) letting the year 1850 adjusted wetland % impact what we calculate in 1851.

Here is my thinking in more detail:

- [ ] First, to make all of this easier and avoid code duplication: Move the whole block of code after the call to `change_landuse` up through the call to `normalizencheck_landuse` (around lines 700-something in mksurfdat.F90) into the top of `normalizencheck_landuse`: I think we're going to want to redo just about all of this for the landuse timeseries file now, and the parts that aren't needed seem like they won't hurt to redo each year.

- [ ] Right after the call to change_landuse (around line 713), but before the call to normalizencheck_landuse: Set 4 new variables:

  ```Fortran
  pctlak_orig(:) = pctlak(:)
  pctwet_orig(:) = pctwet(:)
  pcturb_orig(:) = pcturb(:)
  pctgla_orig(:) = pctgla(:)
  ```

- [ ] Remove the code I just had you add in the time loop before calling normalizencheck_landuse in that loop, since this will now be done in normalizencheck_landuse according to the above item.

- [ ] Right before calling normalizencheck_landuse in the time loop, reset landunit areas that haven't been read back to their original (pre-adjustment) values:

  ```Fortran
  ! For landunits NOT read each year: reset to their pre-adjustment values in preparation for redoing landunit area normalization
  pctwet(:) = pctwet_orig(:)
  pcturb(:) = pcturb_orig(:)
  pctgla(:) = pctgla_orig(:)
  ```

  (Then will come the existing call to normalizencheck_landuse in the time loop.)

- [ ] To make sure this was done right: generate a surface dataset at f09 or f19 resolution from master and after your code changes, and make sure the two are bit-for-bit. (A quick way to check if two files are bit-for-bit identical is to run `/glade/p/cesmdata/cseg/tools/cime/tools/cprnc/cprnc -m /path/to/file1.nc /path/to/file2.nc > cprnc_report.txt`; near the bottom it should report that the two files are identical if they are, or report how many variables differ.) Probably best to do the same with a landuse timeseries file, too; there we expect some diffs due to the inclusion of dynamic lakes now, but I want to be sure that the only differences are as expected. I can help you with this step if you'd like.

- [ ] An additional helpful test would be to compare PCT_LAKE in the first year of a landuse timeseries file with PCT_LAKE on a surface dataset corresponding to that year; hopefully the two will be identical.


Whew! I know that's a lot, but hopefully most of it is relatively straightforward to do. That said: @Ivanderkelen and/or @ekluzek - **I'd welcome your thoughts on the above, because I don't feel 100% sure that what I'm suggesting is correct, and I sometimes feel that, for every issue I fix with mksurfdata_map, I manage to introduce a new issue.** Also, @Ivanderkelen , let me know if this feels like too much for you and if you'd like help implementing the above. The good thing is that we should be able to test to ensure that we aren't messing up any existing fields.

Reply:
Thanks for writing this out! At first sight, it all dounds logic and straightforward to do. I'll keep an eye on https://github.com/ESCOMP/CTSM/pull/1073 and start when it is merged to master.

Reply:
The prosed changes are applied in d8b85afb2d0fd11251e434c5f57a151489bfe05f. 
Maybe good to take a look at the comments in the code, they might be insufficient. 

I had to remove the deallocation of organic, pctclay, pctsand and soicol, as they are rounded now within the normalizencheck_landuse subroutine, and thus repeated in the time loop. (which I think doesn't harm, as it would be more consistent to keep them in the subroutine rather than putting them seperate from rouding the other landunits?). 

I did the testing you proposed: both surfacedatasets (before and after changes) are bit for bit the same. Also, PCT_LAKE from surfdat and first timestep of landuse.timeseries are exactly the same. The differences in the two landuse.timeseries files are due to the rounding (all between -0.5 and 0.5). 

I am not sure yet how important this is to get more detailed dynamics than 0.5%, and will have a look at this tomorrow.

Reply:
Thanks so much @Ivanderkelen . I took a look at your most recent changes and they look great.